NAS-122619 / fs/cifs - add ZFS ACL support to SMB client

anodos325 opened the pull request:
Add an xattr handler with same namespace as ZFS ACL xattr handler to allow userspace utilities to easily preserve and convert contents of SMB Security Descriptor DACL into native ZFS ACL when ingesting data during migration via SMB client (for example using rsync with the explicit option to preserve the xattr in question).

This PR also adds a new procfs endpoint:
/proc/fs/cifs/zfsacl_configuration_options that can be used to control error handling for cases where we can't convert SID into a Unix ID, and currently also whether we allow setting NT ACL via xattr writes on remote SMB server.

Default behavior is:
* Prevent modifications to remote SMB server ACL
* Fail xattr read if ID mapping fails

anodos325 committed:
fs/cifs - add ZFS ACL support to SMB client

Add an xattr handler with same namespace as ZFS ACL xattr handler
to allow userspace utilities to easily preserve and convert
contents of SMB Security Descriptor DACL into native ZFS ACL
when ingesting data during migration via SMB client (for example
using rsync with the explicit option to preserve the xattr in
question).

This PR also adds a new procfs endpoint:
/proc/fs/cifs/zfsacl_configuration_options that can be used to
control error handling for cases where we can't convert SID
into a Unix ID, and currently also whether we allow setting
NT ACL via xattr writes on remote SMB server.

Default behavior is:
* Prevent modifications to remote SMB server ACL
* Fail xattr read if ID mapping fails

ixhamza commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1680,1056 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+	if (unlikely(psid->num_subauth > SID_MAX_SUB_AUTHORITIES)) {
+		cifs_dbg(FYI, "%s: %u subauthorities is too many!\n",
+			 __func__, psid->num_subauth);
+		return -EIO;
```
`EINVAL` or `ERANGE` maybe?

anodos325 replied:
Good point. I don't like that comment either (copy-pasted check from other area). -EINVAL is probably better choice. Since this eventually gets back to xattr handler ERANGE might confuse user.

I'll probably need to go through and think over / justify errors used.

ixhamza commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1680,1056 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+			 */
+			kfree(sidstr);
+			sidtype = SIDOWNER;
+			goto try_upcall_to_get_id;
```
I am not very much familiar with this code path but should there be a termination condition?

anodos325 replied:
This code path gets followed only when `sidtype` is `SIDGROUP`.

Basically, we try first as a group then retry as a user. The reason for trying as group first is because sane/experienced admins use groups to manage permissions :)

ixhamza replied:
Ah, right. Thanks for the clarification.

anodos325 replied:
No problem. This is also good litmus test for where I need comments ;)

bmeagherix commented on fs/cifs/cifs_debug.c:
```diff
@@ -756,9 +768,14 @@ cifs_proc_clean(void)
 	remove_proc_entry("mount_params", proc_fs_cifs);
 
+#ifdef CONFIG_TRUENAS
+	remove_proc_entry("ZFSACL_Flags", proc_fs_cifs);
```
Presumably this should be 
`remove_proc_entry("zfsacl_configuration_flags", proc_fs_cifs);`

(i.e. the parameter supplied to `proc_create` should match the one supplied to `remove_proc_entry`)

Some of the error messages below use the string "ZFSACL_Flags" - it would be better to be consistent.

bmeagherix commented on fs/cifs/cifs_debug.c:
```diff
@@ -1007,6 +1024,69 @@ static const struct proc_ops cifs_security_flags_proc_ops = {
+	unsigned int flags, idmap_flags;
+	char flags_string[12] = { 0 };
+
+	if (copy_from_user(flags_string, buffer, count))
```
Need to check that `count <= 11` ?

bmeagherix commented on fs/cifs/nfs41acl_xdr.h:
```diff
@@ -0,0 +1,125 @@
+#define ACL4_DEFAULTED 0x00000004
+#define ACL4_ISDIR 0x00020000
+
+#define ACE4_ACCESS_ALLOWED_ACE_TYPE 0x0000
+#define ACE4_ACCESS_DENIED_ACE_TYPE 0x0001
+#define ACE4_SYSTEM_AUDIT_ACE_TYPE 0x0002
+#define ACE4_SYSTEM_ALARM_ACE_TYPE 0x0003
```
Maybe add comment line above this block referring to RFC 5661 Section 6.2.1.1

bmeagherix commented on fs/cifs/nfs41acl_xdr.h:
```diff
@@ -0,0 +1,125 @@
+#define NA_ACCESS_MASK_OFFSET 3
+#define NA_WHO_OFFSET 4
+
+#define ACE4_READ_DATA 0x00000001
```
Should there be a comment here referring to RFC 5561 [6.2.1.3](https://datatracker.ietf.org/doc/html/rfc5661#section-6.2.1.3).  ACE Access Mask

bmeagherix commented on fs/cifs/nfs41acl_xdr.h:
```diff
@@ -0,0 +1,125 @@
+
+#define ACE4_ALL_PERMS (ACE4_MODIFY_PERMS|ACE4_WRITE_ACL|ACE4_WRITE_OWNER)
+
+#define ACE4_FILE_INHERIT_ACE 0x00000001
```
Should there be a comment here referring to RFC 5661 [6.2.1.4](https://datatracker.ietf.org/doc/html/rfc5661#section-6.2.1.4).  ACE flag

bmeagherix commented on fs/cifs/nfs41acl_xdr.h:
```diff
@@ -0,0 +1,125 @@
+#define NACE41_LEN 5
+#define NACL_OFFSET 2
+
+#define ACL4_AUTO_INHERIT 0x00000001
+#define ACL4_PROTECTED 0x00000002
+#define ACL4_DEFAULTED 0x00000004
```
Comment with reference to RFC 5661 [6.4.3.2](https://datatracker.ietf.org/doc/html/rfc5661#section-6.4.3.2).  Automatic Inheritance ?

bmeagherix reviewed (changes requested):
Initial/partial review (will do another).

mgrimesix commented on fs/cifs/cifs_debug.c:
```diff
@@ -1007,6 +1024,69 @@ static const struct proc_ops cifs_security_flags_proc_ops = {
+	}
+
+	if (flags & ~MODFLAG_ALL) {
+		cifs_dbg(VFS, "Invalid ZFSACL_Flags: %s\n", flags_string);
```
Maybe give this and the previous error message different text?  e.g. for kstrtouint "Invalid ZFSACL_Flags conversion:" and for this "Invalid ZFSACL_Flags setting:"

mgrimesix commented on fs/cifs/cifs_debug.c:
```diff
@@ -1007,6 +1024,69 @@ static const struct proc_ops cifs_security_flags_proc_ops = {
+	    (idmap_flags == (MODFLAG_FAIL_UNKNOWN_SID | MODFLAG_SKIP_UNKNOWN_SID)) ||
+	    (idmap_flags == (MODFLAG_FAIL_UNKNOWN_SID | MODFLAG_MAP_UNKNOWN_SID)) ||
+	    (idmap_flags == (MODFLAG_SKIP_UNKNOWN_SID | MODFLAG_MAP_UNKNOWN_SID))) {
+		cifs_dbg(VFS, "Only one idmap-related flag may be set");
```
Maybe include the value, similar to the other error messages

mgrimesix commented on fs/cifs/xattr.c:
```diff
@@ -201,6 +205,44 @@ static int cifs_xattr_set(const struct xattr_handler *handler,
+
+		break;
+	}
+#endif /* CONFIG_TRUENASE */
```
Tiny typo in comment

anodos325 committed:
Address first round of review

anodos325 committed:
Add ZFS ACL to listxattr output as needed

bmeagherix commented on fs/cifs/xattr.c:
```diff
@@ -18,13 +18,17 @@
 #define CIFS_XATTR_CIFS_NTSD_FULL "system.cifs_ntsd_full" /* owner/DACL/SACL */
 #define CIFS_XATTR_ATTRIB "cifs.dosattrib"  /* full name: user.cifs.dosattrib */
 #define CIFS_XATTR_CREATETIME "cifs.creationtime"  /* user.cifs.creationtime */
+#define CIFS_XATTR_ZFS_ACL "system.nfs4_acl_xdr" /* DACL only */
```
Should this be inside a `#ifdef CONFIG_TRUENAS ... #endif`

Not essential, but might be cleaner.

bmeagherix commented on fs/cifs/xattr.c:
```diff
@@ -38,7 +42,7 @@
 
 enum { XATTR_USER, XATTR_CIFS_ACL, XATTR_ACL_ACCESS, XATTR_ACL_DEFAULT,
-	XATTR_CIFS_NTSD, XATTR_CIFS_NTSD_FULL };
+	XATTR_CIFS_NTSD, XATTR_CIFS_NTSD_FULL, XATTR_ZFSACL };
```
Likewise to above, should this be inside a `#ifdef CONFIG_TRUENAS ... #endif`

anodos325 committed:
Improve debug messages

anodos325 committed:
Address feedback

bmeagherix commented on fs/cifs/nfs41acl_xdr.h:
```diff
@@ -0,0 +1,151 @@
+
+/*
+ * Native ZFS NFSv41-style ACL is packed in xattr as
+ * follows:
```
Perhaps add "using network byte order", i.e:

> Native ZFS NFSv41-style ACL is packed (using network byte order) in xattr as follows:

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+
+	zfsacl = xdrbuf;
+	*zfsacl++ = 0;
+	*zfsacl++ = htonl(1);
```
More comments for clarity?
```
	*zfsacl++ = 0; /* acl_flags */
	*zfsacl++ = htonl(1); /* ace_count */
```

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+		*nfs_ace_type_out = ACE4_ACCESS_ALLOWED_ACE_TYPE;
+		break;
+	case ACCESS_DENIED_ACE_TYPE:
+		*nfs_ace_type_out = ACCESS_DENIED_ACE_TYPE;
```
Should this not be
`*nfs_ace_type_out = ACE4_ACCESS_DENIED_ACE_TYPE;
`

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+	 * in this case (which will result in upcall to winbind).
+	 */
+	if (psid->num_subauth == 2) {
+		// Possibly Samba encoding of unix id
```
I may be a little hazy on this, so perhaps some further docs would clarify.

Should we also be checking the authority?  (i.e. to make sure it's S-1-22-1-\<UID\> or S-1-22-2-\<GID\>)

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+			return 1;
+		}
+	} else if (psid->num_subauth == 3) {
+		if (psid->sub_auth[1] == sid_unix_NFS_groups.sub_auth[1]) {
```
Likewise, do we also need to check for the 5-88 in S-1-**5-88**-1-\<UID\> or S-1-**5-88**-2-\GID\> ?

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+	error = convert_smb_sid_to_nfs_who(&pace->sid, p_iflag, p_who_id, p_flags);
+	if (error == -ENOKEY) {
+		if (*p_acp_type == ACE4_ACCESS_DENIED_ACE_TYPE) {
+			sid_str = sid_to_key_str(&pace->sid, SIDOWNER);
```
Not seeing where sid_str is declared.  Should this be a local variable here?
`			char *sid_str = sid_to_key_str(&pace->sid, SIDOWNER);
`

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+			kfree(sid_str);
+			return -ENOKEY;
+                }
+		if (global_zfsaclflags & MODFLAG_SKIP_UNKNOWN_SID) {
+			return -EAGAIN;
+		} else if (global_zfsaclflags & MODFLAG_MAP_UNKNOWN_SID) {
+			*p_who_id = from_kuid(&init_user_ns, current_fsuid());
+			return 0;
+		}
```
Since there is no handling required here for `MODFLAG_FAIL_UNKNOWN_SID` (i.e. just falls thru to return the `-ENOKEY`) then maybe add a comment to that effect.

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+	if (pntsd == NULL)
+		return -EIO;
+
+	dacloffset = le32_to_cpu(pntsd->dacloffset);
```
Guard with
```
if (!dacloffset) {
    return generate_null_zfsacl(buf_out);
}
```
?

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+		cifs_dbg(VFS, "%s: ACL size [%u] encoded in NT DACL "
+			 "is invalid.\n",
+			 __func__, le16_to_cpu(dacl_ptr->size));
+		return EINVAL;
```
presumably should be
`return -EINVAL;`

bmeagherix commented on fs/cifs/nfs41acl_xdr.h:
```diff
@@ -0,0 +1,151 @@
+#define ACE4_SPECIAL_GROUP 2
+#define ACE4_SPECIAL_EVERYONE 3
+#define NACE41_LEN 5
+#define NACL_OFFSET 2
```
Is this 2 in order to skip over `acl_flags` and `ace_count`?  If so, maybe add a comment that effect, and/or change name to something more descriptive (e.g. `NACES_OFFSET`).

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+	to_check = flags & ACE4_IDENTIFIER_GROUP ? group : owner;
+
+	/*
+	 * This is a Samba server implementation detail for NFS4 ACL.
+	 * S-1-3-0 and S-1-3-1 are only valid with INHERIT_ONLY set
+	 * whereas owner@ and group@ in NFS4 ACL carry no such restriction.
+	 * Therefore the server will split owner@ into two separate aces:
+	 * one with S-1-3-0 (or S-1-3-1 in case of group@) and INHERIT_ONLY
+	 * and the other as a normal non-special entry for the ID of the user
+	 * or group with no inheritance flags set.
+	 *
+	 * The SMB server will always present the next ACE as the second of
+	 * the pair and so we peek ahead here. If both halves of pair are
+	 * present, then we combine into a single owner@ or group@ entry.
```
Is `isdir` supposed to play a part in this "will split owner@ into two separate aces"?  Code in `calculate_ntsd_acecnt` would also suggest as much.

If so, please include a comment to that effect.

(If not, then does the code below look correct?)

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+	if (!xdr_base)
+		return -ENOMEM;
+
+	zfsacl = (u32 *)xdr_base + NACL_OFFSET;
```
General comment.  In some places `zfsacl` points to the whole (`struct nfsacl4`) buffer, in others (e.g. here & parameter to `calculate_ntsd_acecnt`) it points to the array of aces.  IMO doesn't help clarity.  OTOH `*zfsacl++ = foo` can be a useful mechanism, so I'm not sure I have a suggestion for a fix.

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+		}
+	}
+
+	secdesclen = dacl_ace_cnt * sizeof(struct cifs_ace);
+	secdesclen = max_t(u32, secdesclen, DEFAULT_SEC_DESC_LEN);
+	secdesclen += sizeof(struct cifs_ntsd);
```
AFAICS DEFAULT_SEC_DESC_LEN already includes the `struct cifs_ntsd`.  From _fs/cifs/cifsacl.h_

```
/*
 * Security Descriptor length containing DACL with 3 ACEs (one each for
 * owner, group and world).
 */
#define DEFAULT_SEC_DESC_LEN (sizeof(struct cifs_ntsd) + \
			      sizeof(struct cifs_acl) + \
			      (sizeof(struct cifs_ace) * 4))
```

So perhaps we want to reorder as:
```
secdesclen = dacl_ace_cnt * sizeof(struct cifs_ace);
secdesclen += sizeof(struct cifs_ntsd);
secdesclen = max_t(u32, secdesclen, DEFAULT_SEC_DESC_LEN);

```

I'll also make this comment to make sure I'm understanding correctly.

It seems like we're allocating memory for `dacl_ace_cnt * sizeof(struct cifs_ace)` where `struct cifs_ace` includes space for 15 aub-auth, but then later we only write the number of sub-auth _actually_ in use, and furthermore we pack the SIDs tightly.  Not a problem AFAIK, but ... huh.

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+}
+
+/*
+ * This is special handling for either NULL or emtpy ACLs.
```
emtpy => empty

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+	uint sidtype = flags & ACE4_IDENTIFIER_GROUP;
+	uid_t id;
+
+	if (iflag == 0) {
```
Maybe add comment:
`    if (iflag == 0) { /* Not ACEI4_SPECIAL_WHO */`

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+	 * an identical access mask. One is non-inheriting for the inode owner
+	 * or group, and the other is inherit-only with the special SID value.
+	 */
+	if (iflag && (who_id != ACE4_SPECIAL_EVERYONE) &&
```
Would it be slightly clearer (and more future-proof) if we wrote
`	if ((iflag & ACEI4_SPECIAL_WHO) && (who_id != ACE4_SPECIAL_EVERYONE) &&`

(If so does that mean it would be clearer/more future-proof to use this pattern _everywhere_ where `iflag` is tested for non-zero?)

"meh" is a valid response!

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+	return -EINVAL;
+}
+
+#define BASE_ACE_SIZE (1 + 1 + 2 + 4)
```
Maybe add a comment to make clear that this corresponds to `struct cifs_ace`
`#define BASE_ACE_SIZE (1 + 1 + 2 + 4) /* struct cifs_ace: type, flags, size, access_req */`

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+		return 0;
+	}
+
+	pnntsd->dacloffset = cpu_to_le32(sizeof(struct cifs_ntsd));
+	dacl = (struct cifs_acl *)(pnntsd + sizeof(struct cifs_ntsd));
+	dacl->size = cpu_to_le16(sizeof(struct cifs_acl));
+	dacl->revision = cpu_to_le16(ACL_REVISION);
```
This is the `dacl_is_empty`, correct?  If so, add comment:
`/* dacl_is_empty */`

bmeagherix commented on fs/cifs/cifsacl.c:
```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+		kfree(pnntsd);
+	}
+
+	*ppntsd_out = pnntsd;
+	*acllen_out = secdesclen;
```
Probably better not to set these if we're returning an error ?

bmeagherix reviewed (changes requested):
Mostly tidy/clarification suggestions, but a few possible code changes.

anodos325 committed:
fix typo

"this is just a small change, I don't need to check it locally"
he said before pushing into the remote git branch.

anodos325 committed:
Address more feedback

anodos325 committed:
Fix sign of error return
